text/0000-mergeable-rustdoc-cross-crate-info.md

```diff
@@ -0,0 +1,441 @@
+
+# Guide-level explanation
+
+In this example, there is a crate `t` which defines a trait `T`, and a crate `s` which defines a struct `S` that implements `T`. Our goal in this demo is for `S` to appear as in implementer in `T`'s docs, even if `s` and `t` are documented independently. This guide will be assuming that we want a crate `i` that serves as our documentation index. See the Unresolved questions section for ideas that do not require an index crate.
+
+```shell
+mkdir -p t/src s/src i/src merged/doc
+echo "pub trait T {}" > t/src/lib.rs
+echo "pub struct S; impl t::T for S {}" > s/src/lib.rs
+MERGED=file://$(realpath merged/doc)
+```
+
+[Actively use](https://doc.rust-lang.org/rustc/command-line-arguments.html#--extern-specify-where-an-external-library-is-located) `t` and `s` in `i`. The `extern crate` declarations are not needed if the crates are otherwise referenced in the index; intra-doc links are enough.
+
+```shell
+echo "extern crate t; extern crate s;" > i/src/lib.rs
+```
+
+Compile the crates.
+
+```shell
+rustc --crate-name=t --crate-type=lib --edition=2021 --emit=metadata --out-dir=t/target t/src/lib.rs
+rustc --crate-name=s --crate-type=lib --edition=2021 --emit=metadata --out-dir=s/target --extern t=t/target/libt.rmeta s/src/lib.rs
+rustc --crate-name=i --crate-type=lib --edition=2021 --emit=metadata --out-dir=i/target --extern s=s/target/libs.rmeta --extern t=t/target/libt.rmeta -L t/target i/src/lib.rs
+```
+
+Document `s` and `t` independently, providing `--merge=none` and `--write-info-json=<crate name>/target/doc.parts`
+
+```shell
+rustdoc -Z unstable-options --crate-name=t --crate-type=lib --edition=2021 --out-dir=t/target/doc --extern-html-root-url t=$(MERGED) --merge=none --write-info-json=t/target/doc.parts t/src/lib.rs
+rustdoc -Z unstable-options --crate-name=s --crate-type=lib --edition=2021 --out-dir=s/target/doc --extern-html-root-url s=$(MERGED) --extern-html-root-url t=$(MERGED) --merge=none --write-info-json=t/target/doc.parts --extern t=t/target/libt.rmeta s/src/lib.rs
+```
+
+Link everything with a final invocation of rustdoc on `i`. We will provide `--merge=write-only` and `--include-info-json=<path>`. See the Reference-level explanation about these flags.
+
+```shell
+rustdoc -Z unstable-options --crate-name=i --crate-type=lib --edition=2021 --enable-index-page --out-dir=i/target/doc --extern-html-root-url s=$(MERGED) --extern-html-root-url t=$(MERGED) --extern-html-root-url i=$(MERGED) --merge=write-only --include-info-json t=t/target/doc.parts --include-info-json s=s/target/doc.parts --extern t=t/target/libt.rmeta --extern s=s/target/libs.rmeta -L t/target i/src/lib.rs
+```
+
+Merge the docs with `cp`. This can be avoided if `--out-dir=$(MERGED)` is used for all of the rustdoc calls. We copy here to illustrate that documenting `s` is independent of documenting `t`, and could happen on separate machines.
```

Comment:
`cp` is the stand-in for the step that takes docs from external crates and collects them in a single web root. Depending on the location of the external crates, this step may look very different:
* It may do nothing; `Cargo` generates the docs in a single `doc/` folder.
* `cp` or `mv` files from one location on the filesystem to another. This is shown in the example in the RFC.
* `scp`, `rsync`, `wget --recursive`, a proprietary protocol, etc. This RFC is intended to enable CCI in crates that have been documented on separate machines.

Because of the wide variety of locations that external docs may be, I believe that the build system is best positioned to collect them into a single location.

When you pass the `--merge=none` flag to document the external crates, the shared resources are not generated. Only `crate-info.json` and the item docs will be generated for these crates. The shared resources are only generated when you pass `--merge=auto` (default) or `--merge=write-only`, which is only provided when documenting the index or root crate. There are not any redundant files to bloat the disk space.

Reply:
Then please update so that it is mentioned that rustdoc will handle the copying internally as we don't want to rely on external tools.

Reply:
Okay, I will update it so that rustdoc does the copying.